docs/alert_smtp_filtered.5.md
alert_smtp_filtered.sh 5 "July 2025" alert_smtp_filtered.sh "User Manual"
==================================================

# NAME

alert_smtp_filtered.sh - Sample SMTP alert agent for pacemaker with filtering

# SYNOPSIS

pcs alert create id=filtered-smtp \
 path=/var/lib/pacemaker/**alert_smtp_filtered.sh** options \
 **email_sender=**noreply@example.com **RHA_alert_kind=**"fencing"


# DESCRIPTION

**alert_smtp_filtered.sh** is a sample alert agent which implements filtering by
 matching the value of pacemaker's CRM_alert_kind variable that is set when an 
 alert is generated. This agent was built for a client who wished to send 
 receive alerts whenever resources are relocated,.

By default, the email client the script expects is sendmail.

# OPTIONS

**path=**_/var/lib/pacemaker/alert_smtp_filtered.sh_

This is the path to the alert agent on the nodes' filesystems - the path should
 match to wherever you've installed the file. By default they're placed in 
 /usr/share/pacemaker/alerts/ when installing from rpm, and they're usually 
 manually placed in /var/lib/pacemaker/ for runtime when the agents are 
 configured.

**email_sender=**_user@example.com_

This is what the agent will use as the "FROM" field on email alerts

**RHA_alert_kind=**_"fencing,node,resource"_

This option sets the RHA_alert_kind variable in the alert_smtp_filtered alert 
 agent, to specify the criteria on which alerts to allow to send to the email 
 recipient.

Note that otherwise-unspecified alert types will be sent to the recipient 
 regardless of the filter specification.

_fencing_

These alerts are generated when a node is fenced, whether automatically or 
 automatically.

_node_

These alerts when a node is suspended, unsuspended, rebooted, joins the 
 cluster, etc. 
    
_resource_

These alerts are generated when a resource is started, stopped, or fails to 
 start. 
    
# INSTALLATION

Place alert_syslog.sh in pacemaker lib dirctory (typically /var/lib/pacemaker)
 chown it the pacemaker user and group (typically hacluster:haclient on a 
 default install); chmod it 0750

```
[root@nodea ~]# cp /usr/share/pacemaker/alerts/alert_smtp_filtered.sh.sample \
  /var/lib/pacemaker/alert_smtp_filtered.sh
[root@nodea ~]# chown hacluster:haclient /var/lib/pacemaker/alert_smtp_filtered.sh
[root@nodea ~]# chmod 0750 /var/lib/pacemaker/alert_smtp_filtered.sh
```

Proceed to EXAMPLES section for alert configuration

# EXAMPLES

The following example will send alert emails whenever a node is fenced and also 
 unhandled alerts, but not node or resource alerts, and send the 
 alert emails to sysad@example.com

```
[root@nodea ~]# pcs alert create id=filtered-smtp \
 path=/var/lib/pacemaker/alert_smtp_filtered.sh options \
 email_sender=noreply@example.com RHA_alert_kind="fencing"
[root@nodea ~]# pcs alert recipient add filtered-smtp \
 value=sysad@example.com@example.com  
[root@nodea ~]# 
```

This example will send alerts of kind node, resource, and "unhandled"
 alerts, but not fencing notifications, and the alert emails will go to 
 monitor@example.com:

```
[root@nodea ~]# pcs alert create id=filtered-smtp \
path=/var/lib/pacemaker/alert_smtp_filtered.sh options \
email_sender=noreply@example.com RHA_alert_kind="node,resource"
[root@nodea ~]# pcs alert recipient add filtered-smtp \
value=monitor@example.com  
[root@nodea ~]# 
```

# HISTORY
July 2025, Originally compiled by Kimberly Lazarski (klazarsk@redhat.com)


agents/alert_smtp_filtered.sh
#!/bin/sh
#
# Adapted from alert_smtp.sh from the pacemaker package by 
# klazarsk@redhat.com for scopable alert generation
#
# You can improve upon the granularity of the alerting by matching 
# substrings in variables such as CRM_alert_desc and testing other
# variables and creating more complex cases to drive filtering and turning 
# individual alerts off and on. In this user's case, they only wanted 
# fencing notices and wanted it filtered at the the alert generation stage.
# 
#
##############################################################################
#
# Copyright 2016-2021 the Pacemaker project contributors
#
# The version control history for this file may have further details.
#
# This source code is licensed under the GNU General Public License version 2
# or later (GPLv2+) WITHOUT ANY WARRANTY.
#
##############################################################################
#
# Sample configuration (cib fragment in xml notation)
# ================================
# <configuration>
#   <alerts>
#     <alert id="smtp_alert" path="/path/to/alert_smtp">
#       <instance_attributes id="config_for_alert_smtp">
#         <nvpair id="cluster_name" name="cluster_name" value=""/>
#         <nvpair id="email_client" name="email_client" value=""/>
#         <nvpair id="email_sender" name="email_sender" value=""/>
#       </instance_attributes>
#       <recipient id="smtp_destination" value="admin@example.com"/>
#     </alert>
#   </alerts>
# </configuration>

# Explicitly list all environment variables used, to make static analysis happy
: ${CRM_alert_version:=""}
: ${CRM_alert_recipient:=""}
: ${CRM_alert_timestamp:=""}
: ${CRM_alert_kind:=""}
: ${CRM_alert_node:=""}
: ${CRM_alert_desc:=""}
: ${CRM_alert_task:=""}
: ${CRM_alert_rsc:=""}
: ${CRM_alert_attribute_name:=""}
: ${CRM_alert_attribute_value:=""}

#############################################################
# Pass "RHA_alert_kinds" as an option at the pcs alert create
# stage, otherwise if the variable is null/not set by alert 
# options assignment, take the value in this stanza
# Take alert types out of optAlertKinds to disable alerts
  if [ -z ${RHA_alert_kinds} ]; then
  
    # ALL alerts (unfiltered)
    optAlertKinds="fencing,node,resource"
    # ONLY fencing alerts:
    # optAlertKinds="fencing"
  
  else
   
    optAlertKinds="${RHA_alert_kinds}"
  
  fi 
#
#############################################################


email_client_default="sendmail"
email_sender_default="hacluster"
email_recipient_default="root"

: ${email_client=${email_client_default}}
: ${email_sender=${email_sender_default}}
email_recipient="${CRM_alert_recipient-${email_recipient_default}}"

node_name=$(uname -n)
cluster_name=$(crm_attribute --query -n cluster-name -q)
email_body=$(env | grep CRM_alert_ ; env | grep RHA_[as]; echo -e "\n note: Email notifications limited to alert types: ${optAlertKinds}\n")

if [ ! -z "${email_sender##*@*}" ]; then
    email_sender="${email_sender}@${node_name}"
fi

if [ ! -z "${email_recipient##*@*}" ]; then
    email_recipient="${email_recipient}@${node_name}"
fi

if [ -z ${CRM_alert_version} ]; then
    email_subject="Pacemaker version 1.1.15 or later is required for alerts"
else
  case ${CRM_alert_kind} in
    node)
      if [[ $optAlertKinds == *"node"* ]]; then
        email_subject="${CRM_alert_timestamp} ${cluster_name}: Node '${CRM_alert_node}' is now '${CRM_alert_desc}'"
      fi
      ;;
    fencing)
      if [[ $optAlertKinds == *"fencing"* ]]; then
        email_subject="${CRM_alert_timestamp} ${cluster_name}: Fencing ${CRM_alert_desc}"
      fi
      ;;
    resource)
      if [[ $optAlertKinds == *"resource"* ]]; then
        if [ ${CRM_alert_interval} = "0" ]; then
            CRM_alert_interval=""
        else
            CRM_alert_interval=" (${CRM_alert_interval})"
        fi

        if [ ${CRM_alert_target_rc} = "0" ]; then
            CRM_alert_target_rc=""
        else
            CRM_alert_target_rc=" (target: ${CRM_alert_target_rc})"
        fi

        case ${CRM_alert_desc} in
          Cancelled) ;;
          *)
              email_subject="${CRM_alert_timestamp} ${cluster_name}: Resource operation '${CRM_alert_task}${CRM_alert_interval}' for '${CRM_alert_rsc}' on '${CRM_alert_node}': ${CRM_alert_desc}${CRM_alert_target_rc}"
          ;;
        esac
      fi
      ;;
    *)
        email_subject="${CRM_alert_timestamp} ${cluster_name}: Unhandled $CRM_alert_kind alert"
        ;;

  esac
fi

if [ ! -z "${email_subject}" ]; then
    case $email_client in
        # This sample script supports only sendmail for sending the email.
	# Support for additional senders can easily be added by adding
	# new cases here.
        sendmail)
            sendmail -t -r "${email_sender}" <<__EOF__
From: ${email_sender}
To: ${email_recipient}
Return-Path: ${email_sender}
Subject: ${email_subject}

${email_body}
__EOF__
            ;;
        *)
            ;;
    esac
fi
